src/magic_spell_effect.cpp

```diff
@@ -503,3 +509,245 @@ void spell_effect::translocate( const spell &sp, const Creature &caster,
+        monster *const mon = g->critter_at<monster>( dest, true );
+        if( mon && mon->friendly == 0 && rng( 0, 600 ) > mon->get_hp() ) {
+            mon->make_friendly();
+        }
+    }
+}
+
+void spell_effect::bugs( const spell &sp, const Creature &caster, const tripoint &target )
+{
+    int roll = rng( 1, 10 );
+    mtype_id bug = mtype_id::NULL_ID();
+    int num = 0;
+    std::vector<tripoint> empty;
+    for( const tripoint &dest : g->m.points_in_radius( target, sp.aoe() ) ) {
+        if( g->is_empty( dest ) ) {
+            empty.push_back( dest );
+        }
+    }
+    if( empty.empty() || roll <= 4 ) {
+        caster.add_msg_if_player( m_warning, _( "Flies buzz around you." ) );
+    } else if( roll <= 7 ) {
+        caster.add_msg_if_player( m_warning, _( "Giant flies appear!" ) );
+        bug = static_cast<mtype_id>( "mon_fly" );
+        num = rng( 2, 4 );
+    } else if( roll <= 9 ) {
+        caster.add_msg_if_player( m_warning, _( "Giant bees appear!" ) );
+        bug = static_cast<mtype_id>( "mon_bee" );
+        num = rng( 1, 3 );
+    } else {
+        caster.add_msg_if_player( m_warning, _( "Giant wasps appear!" ) );
+        bug = static_cast<mtype_id>( "mon_wasp" );
+        num = rng( 1, 2 );
+    }
+    if( bug ) {
+        for( int j = 0; j < num && !empty.empty(); j++ ) {
+            const tripoint spawnp = random_entry_removed( empty );
+            if( monster *const b = g->summon_mon( bug, spawnp ) ) {
+                b->friendly = -1;
+                b->add_effect( static_cast<efftype_id>( "pet" ), 1_turns, num_bp, true );
+            }
+        }
+    }
+}
+
+void spell_effect::light( const spell &sp, const Creature &caster )
+{
+    caster.add_msg_if_player( _( "The %s glows brightly!" ), sp.get_obj_name() );
+    g->events.add( EVENT_ARTIFACT_LIGHT, calendar::turn + 3_minutes );
+}
+
+void spell_effect::growth( const tripoint &target )
+{
+    monster tmptriffid( mtype_id::NULL_ID(), target );
+    mattack::growplants( &tmptriffid );
+}
+
+void spell_effect::mutate( const Creature &caster )
+{
+    player *p = g->critter_at<player>( caster.pos() );
+    if( !one_in( 3 ) ) {
+        p->mutate();
+    }
+}
+
+void spell_effect::teleglow( const Creature &caster )
+{
+    player *p = g->critter_at<player>( caster.pos() );
+    p->add_msg_if_player( m_warning, _( "You feel unhinged." ) );
+    p->add_effect( efftype_id( "effect_teleglow" ), rng( 30_minutes, 120_minutes ) );
+}
+
+void spell_effect::noise( const spell &sp, const Creature &caster, const tripoint &target )
+{
+    sounds::sound( target, sp.damage(), sounds::sound_t::combat,
+                   string_format( _( "a deafening boom from %s %s" ),
+                                  caster.disp_name( true ), sp.get_obj_name() ), true, "misc", "shockwave" );
+}
+
+void spell_effect::scream( const spell &sp, const Creature &caster, const tripoint &target )
+{
+    player *p = g->critter_at<player>( caster.pos() );
```

Comment:
```suggestion
    player &p = dynamic_cast<player&>( caster );
```
And change the parameter to be non-const.

Reply:
thx. Ill keep it in mind. This function was refractured.